### PR TITLE
Typegen

### PR DESCRIPTION
**Pull request by chelouche9**

### CODE REVIEW

**Comment by chelouche9** [comment]
@xenova this is only the utils.js file. It will take some time lol. 
I also think we might add a prettier and linter so people won't override all file like me 😝.

File: src/utils.js [hunk 1]
```diff
@@ -207,185 +286,279 @@ async function fetchJSON(modelPath, fileName, progressCallback = null, fatal = t
...
+ */
 function getTopItems(items, top_k = 0) {
-    // if top == 0, return all
+  // if top == 0, return all
 
-    items = Array.from(items)
-        .map((x, i) => [i, x])            // Get indices ([index, score])
-        .sort((a, b) => b[1] - a[1])      // Sort by log probabilities
+  items = Array.from(items)
+    .map((x, i) => [i, x]) // Get indices ([index, score])
```
Comment (kungfooman) [inline src/utils.js:463]: I would rather have the comments aligned and it might be best to focus on JSDoc only in one PR, otherwise it becomes hard to judge

File: src/utils.js [hunk 2]
```diff
@@ -207,185 +286,279 @@ async function fetchJSON(modelPath, fileName, progressCallback = null, fatal = t
...
-        return Object.setPrototypeOf(closure, new.target.prototype)
-    }
-
-    _call(...args) {
-        throw Error('Must implement _call method in subclass')
-    }
+  /**
+   * Creates a new instance of the Callable class.
+   *
+   * @constructor
```
Comment (kungfooman) [inline src/utils.js:525]: not needed

File: src/utils.js [hunk 3]
```diff
@@ -207,185 +286,279 @@ async function fetchJSON(modelPath, fileName, progressCallback = null, fatal = t
...
+      return closure._call(...args);
+    };
+    return Object.setPrototypeOf(closure, new.target.prototype);
+  }
+
+  /**
+   * This method should be implemented in subclasses to provide the
+   * functionality of the callable object.
+   *
+   * @method _call
```
Comment (kungfooman) [inline src/utils.js:538]: not needed

File: src/utils.js [hunk 4]
```diff
@@ -207,185 +286,279 @@ async function fetchJSON(modelPath, fileName, progressCallback = null, fatal = t
...
+   * @throws {Error} Must implement _call method in subclass
+   */
+  _call(...args) {
```
Comment (kungfooman) [inline src/utils.js:541]: args param is never defined

File: src/utils.js [hunk 5]
```diff
@@ -207,185 +286,279 @@ async function fetchJSON(modelPath, fileName, progressCallback = null, fatal = t
...
 
-
+/**
+ * Check if a value is an integer.
+ * @param {*} x - The value to check.
+ * @returns {boolean} - True if the value is a string, false otherwise.
+ */
 function isIntegralNumber(x) {
-    return Number.isInteger(x) || typeof x === 'bigint'
+  return Number.isInteger(x) || typeof x === "bigint";
```
Comment (kungfooman) [inline src/utils.js:561]: Why bother replacing every ' with "?

Comment (chelouche9) [reply src/utils.js:561]: It's my prettier. Forgot to disable it.

**Comment by xenova** [comment]
I just resolved a conflict coming from main. Let me know when you want me to review and test fully 👍

File: src/tensor_utils.js [hunk 6]
```diff
@@ -59,7 +74,12 @@ class Tensor extends ONNX.Tensor {
     // TODO add .slice()
 }
 
-
+/**
+ * Reshapes a 1-dimensional array into an n-dimensional array, according to the provided dimensions.
+ * @param {Array} data - The input array to reshape.
+ * @param {Array} dimensions - The target shape/dimensions.
+ * @returns {Array} The reshaped array.
+ */
```
[the remark this answers is not in the document]
Comment (xenova) [reply src/tensor_utils.js:82]: That's some big brain stuff right there.

[the remark this answers is not in the document]
Comment (xenova) [reply src/tensor_utils.js]: Sounds good! Thanks again!

**Comment by xenova** [comment]
> > I just resolved a conflict coming from main. Let me know when you want me to review and test fully 👍
> 
> @xenova I have left only 3 files. Hopefully in two days.

Sweet! No rush :)

**Comment by chelouche9** [comment]
@xenova The PR is ready, all files have JSdoc now :)

**Comment by xenova** [comment]
Nice! 🎉 I'm almost done adding a few more models/tasks to main, then I'll review, merge, and finally add any missing docs.

**Comment by xenova** [comment]
> Just a general question for you @xenova, do you prefer to keep the source in JavaScript vs. moving some things over to TypeScript?

Not too sure 👀 ... JS has been easier (and faster) to develop in since I don't have much experience with TS. 

But I'm pretty sure that *eventually*, the project will transition to TS.

**Comment by kungfooman** [comment]
> But I'm pretty sure that _eventually_, the project will transition to TS.

I hope not. TS is useful powering a language server, but actually writing TS is ugly once the types become a bit more complicated. And you lose the ability to load code directly as ES6 modules in the browser via `<script type="module">`.

I would urge caution with regards to those who seem overly enthusiastic about TS, as their enthusiasm is often just following "that shiny new thing", while being disconnected from pragmatism.

Simple job for you, @biw, to make things clear, translate this to TypeScript:

```js
/**
 * @example
 * forInMap({a: 11, b: 22}, (k, v) => v); // [11, 22]
 * @param {Record<KEY, VAL>} o - The object.
 * @param {(key: string, value: VAL) => RET} lambda - The function that operates on the key/value pairs.
 * @template {string | number | symbol} KEY  - The key.
 * @template VAL - Type of the value given to `lambda`.
 * @template RET - Type of the return value of `lambda`.
 * @returns {RET[]}
 */
function forInMap(o, lambda) {
  return Object.keys(o).map(key => lambda(key, o[key]));
}
```

File: src/processors.js [hunk 7]
```diff
@@ -336,6 +427,12 @@ class WhisperFeatureExtractor extends FeatureExtractor {
         return frames;
     }
 
+    /**
+     * Generates a Hanning window of length M.
+     *
+     * @param {number} M - The length of the Hanning window to generate.
+     * @returns {Float32Array} - The generated Hanning window.
```
Comment (kungfooman) [inline src/processors.js:434]:
```suggestion
     * @returns {Float32Array | number[]} - The generated Hanning window.
```

**Comment by kungfooman** [comment]
> @xenova The PR is ready, all files have JSdoc now :)

Great work :+1: 

I looked over it and `src/utils.js` received way more changes than necessary. I see the value of a linter, but maybe we have to keep it separate? It looks like it moved methods up and down etc.

**Comment by chelouche9** [comment]
@kungfooman thanks for the feedback. The linter is turned off. I think it happened when resolving conflicts. I will have a look and fix :)

**Comment by chelouche9** [comment]
@xenova just checking what is missing to merge this :)

**Comment by biw** [comment]
> I hope not. TS is useful powering a language server, but actually writing TS is ugly once the types become a bit more complicated. And you lose the ability to load code directly as ES6 modules in the browser via `<script type="module">`.

Since there's already a build pipeline with webpack, this isn't an issue. 😄

> I would urge caution with regards to those who seem overly enthusiastic about TS, as their enthusiasm is often just following "that shiny new thing", while being disconnected from pragmatism.

As someone who's been shipping typescript code in production for over four years, my intentions are not to chase the new shiny thing. Just trying to help out if the need is there. Not trying to force anyone to use anything 😄

> 
> Simple job for you, @biw, to make things clear, translate this to TypeScript:
> 
> ```js
> /**
>  * @example
>  * forInMap({a: 11, b: 22}, (k, v) => v); // [11, 22]
>  * @param {Record<KEY, VAL>} o - The object.
>  * @param {(key: string, value: VAL) => RET} lambda - The function that operates on the key/value pairs.
>  * @template {string | number | symbol} KEY  - The key.
>  * @template VAL - Type of the value given to `lambda`.
>  * @template RET - Type of the return value of `lambda`.
>  * @returns {RET[]}
>  */
> function forInMap(o, lambda) {
>   return Object.keys(o).map(key => lambda(key, o[key]));
> }
> ```

Here you go!
```ts
function forInMap2<V, K extends Record<string | number | symbol, V>, Ret>(
  o: K,
  lambda: (key: keyof K, value: K[typeof key]) => Ret,
) {
  return Object.keys(o).map((key) => lambda(key, o[key as keyof K]))
}
```

Not trying to start a flame war, I just wanted to be helpful if @xenova had an interest in using TypeScript.

**Comment by xenova** [comment]
> @xenova just checking what is missing to merge this :)

Hi! No need to do anything on your side - I will get to fully reviewing (and merging) by the end of the week (hopefully lol).

**Comment by xenova** [comment]
> Not trying to start a flame war, I just wanted to be helpful if @xenova had an interest in using TypeScript.

Haha no worries! 🤣 I'm definitely open to learning new tech, so, we shall see ;) For now, as you'd expect, I'll continue developing in JS, otherwise, I'd probably not be able to keep up with all the craziness in the AI space right now 😄

**Comment by xenova** [comment]
Fixed the conflicts :+1: VSCode is displaying quite a few warnings/errors, and it will probably be good to get those (mostly) resolved before merging. There are also some methods/classes which were added recently that aren't fully documented. If one of you wants to maybe help with these 2 things, that would be great!

Some other exciting news: I've been talking to some HF staff, and they are willing to host the docs on https://huggingface.co/docs/ 🤯 , and then listing Transformers.js as a community-driven open-source project! 🔥

**Comment by jcbhmr** [comment]
Mentioning it here; I used this as a base for a full `.ts` conversion (well, not _full_ full, but getting there! 😉) in https://github.com/jcbhmr/transformers.js/pull/13 
https://github.com/chelouche9/transformers.js/compare/typegen...jcbhmr:transformers.js:use-ts

**Comment by chelouche9** [comment]
> Some other exciting news: I've been talking to some HF staff, and they proposed hosting the docs on https://huggingface.co/docs/ 🤯 , and then listing Transformers.js as a community-driven open-source project! 🔥 Super exciting!

Wow great news! Good job @xenova. Your hard work will reach much more people.

> Fixed the conflicts 👍 VSCode is displaying quite a few warnings/errors, and it will probably be good to get those (mostly) resolved before merging. There are also some methods/classes which were added recently that aren't fully documented. If one of you wants to maybe help with these 2 things, that would be great!

I was off the last couple of days. I will start tomorrow to try and fix some conflicts. Do you have any leads what is the most important part need fixing?

**Comment by xenova** [comment]
> I was off the last couple of days. I will start tomorrow to try and fix some conflicts. Do you have any leads what is the most important part need fixing?

Sounds good! 🔥 Other than some missing JSDocs (after merging main into the PR), it is just some warnings. The tests run successfully, so, there's nothing "fatally" wrong... 

The main thing that pops up is a "Property not defined" error, which is understandable since not all models share the same instance variables (and because of the way I'm doing inheritance to reduce code duplication). For example, encoder-decoder must define some extra properties such as `num_decoder_layers`, `num_decoder_heads`, `decoder_dim_kv`, `num_encoder_layers`, `num_encoder_heads`, and `encoder_dim_kv`. Of course, these don't exist on the base class, so, it is giving a warning saying that we are accessing/setting a property that doesn't exist. Not too sure how to hide/fix that though. If there is a way to globally disable those problems, that could work.

So, other than just adding the missing JSDocs, you could maybe just try fix/reduce some of the warnings (no real particular order tbh). If you use VSCode, you should be able to see the "Problems" (there are quite a lot haha).

**Comment by chelouche9** [comment]
@xenova I rebased over main and there were many conflicts. I hope nothing breaks. 

Merging this PR as soon as possible will help us avoid doing those time-consuming rebase. 
However, we don't want to add all the errors. Since adding types to the code requires a deep understanding of its usage, I suggest that I will change it to any anywhere it breaks and we will incrementally change it over time. 

What do you think?

**Comment by xenova** [comment]
> Merging this PR as soon as possible will help us avoid doing those time-consuming rebase. 

Agreed! I'm trying to get to all open PRs now (with this one being the number 1 priority).

> However, we don't want to add all the errors. Since adding types to the code requires a deep understanding of its usage, I suggest that I will change it to any anywhere it breaks and we will incrementally change it over time.
> 
> What do you think?

Sounds good 👌

**Comment by chelouche9** [comment]
@xenova I have fixed everything that is "type" related. Some errors are due to different js versions (like not using the spread operator below ecma2015), and some are about the _call method on objects. To fix them we need to make some code changes. We can do that on dedicated PRs later. 

There were a lot of conflicts, I made sure everything is correct but I would recommend you test it and give an additional look :)

**Comment by xenova** [comment]
Awesome! 💪 Once again, it is kind of difficult to see where you made the changes, so, could you fix the formatting like you did before? Thanks :)

I'll try get it merged tonight if I can 🥳

**Comment by chelouche9** [comment]
>Awesome! 💪 Once again, it is kind of difficult to see where you made the changes, so, could you fix the formatting like you did before? Thanks :)

@xenova done

File: src/tokenizers.js [hunk 8]
```diff
@@ -173,7 +173,7 @@ class WordPieceTokenizer extends TokenizerModel {
                 start = end;
             }
             if (isUnknown) {
-                outputTokens.push(this.unk_token);
+                outputTokens.push(this.unknownToken);
```
Comment (xenova) [inline src/tokenizers.js]: This should be `unk_token`

**Comment by chelouche9** [comment]
@xenova I made another go. Now all the files are clean from the formatting and the weird code changes.

File: jsconfig 2.json
Comment (kungfooman) [inline jsconfig 2.json:1]: Was it an accident to commit this file?

File: src/fft.js [hunk 9]
```diff
@@ -37,16 +43,35 @@ class FFT {
...
     createComplexArray() {
         return new Float64Array(this._csize);
     }
+
+    /**
+     * Converts a complex number representation stored in a Float64Array to an array of real numbers.
+     * 
+     * @param {Float64Array} complex - The complex number representation to be converted.
+     * @param {Array} [storage] - An optional array to store the result in.
+     * @returns {Array} An array of real numbers representing the input complex number representation.
```
Comment (kungfooman) [inline src/fft.js:60]:
```suggestion
     * @param {number[]} [storage] - An optional array to store the result in.
     * @returns {number[]} An array of real numbers representing the input complex number representation.
```

To be a bit more specific

File: src/pipelines.js [hunk 10]
```diff
@@ -102,7 +137,20 @@ class TextClassificationPipeline extends Pipeline {
...
 class QuestionAnsweringPipeline extends Pipeline {
+    /**
+     * Executes the question answering task.
+     * @param {string|string[]} question - The question(s) to be answered.
+     * @param {string|string[]} context - The context(s) where the answer(s) can be found.
+     * @param {object} options - An optional object containing the following properties:
+     * @param {number} [options.topk=1] - The number of top answer predictions to be returned.
+     * @returns {Promise<any>} - A promise that resolves to an array or object containing the predicted answers and scores.
+     */
+    // TODO: fix error below
```
Comment (kungfooman) [inline src/pipelines.js:153]: if you want to keep the todo, use `@todo` is JSDoc

File: src/pipelines.js [hunk 11]
```diff
@@ -272,11 +362,11 @@ class TextGenerationPipeline extends Pipeline {
             inputs_attention_mask: attention_mask
         });
 
-        let toReturn = outputTokenIds.map((outTokens, i) => {
+        let toReturn = outputTokenIds.map((/** @type {any} */ outTokens, /** @type {string | number} */ i) => {
             let startText = texts[i].trim();
             let decoded = this.tokenizer.batch_decode(outTokens, {
                 skip_special_tokens: true,
-            }).map(x => {
+            }).map((/** @type {any} */ x) => {
```
Comment (kungfooman) [inline src/pipelines.js:369]: Same here, can you declare the return type for `batch_decode`, so `map` can infer the type without any hints?

File: src/pipelines.js [hunk 12]
```diff
@@ -329,7 +434,7 @@ class ZeroShotClassificationPipeline extends Pipeline {
 
         // Insert labels into hypothesis template
         let hypotheses = candidate_labels.map(
-            x => hypothesis_template.replace('{}', x)
+            (            /** @type {string} */ x) => hypothesis_template.replace('{}', x)
```
Comment (kungfooman) [inline src/pipelines.js:437]: Same here, just writing the types over the variables like `/** @type {string[]} */` helps to infer the type automatically helps to remove these type hints

File: src/pipelines.js [hunk 13]
```diff
@@ -531,6 +692,9 @@ class AutomaticSpeechRecognitionPipeline extends Pipeline {
         for (let aud of audio) {
             aud = await this._preprocess(aud, sampling_rate)
 
+            /**
+             * @type {any}
```
Comment (kungfooman) [inline src/pipelines.js:696]:
```suggestion
             * @type {any[]}
```

Could be at least an "any array"?

File: src/pipelines.js [hunk 14]
```diff
@@ -313,6 +412,12 @@ class ZeroShotClassificationPipeline extends Pipeline {
             this.contradiction_id = 0;
         }
     }
+    /**
+     * @param {any[]} texts
+     * @param {any[]} candidate_labels
+     * @return {Promise<*>}
+     */
+    // TODO: fix error below
```
Comment (kungfooman) [inline src/pipelines.js:420]: If you want to keep the todo, use `@todo` is JSDoc

File: src/pipelines.js [hunk 15]
```diff
@@ -674,12 +869,33 @@ class ImageClassificationPipeline extends Pipeline {
...
+
+    /**
+     * Classify the input images with candidate labels using a zero-shot approach.
+     * @param {Array} images - The input images.
+     * @param {Array} candidate_labels - The candidate labels.
+     * @param {Object} options - The options for the classification.
+     * @param {string} [options.hypothesis_template] - The hypothesis template to use for zero-shot classification. Default: "This is a photo of {}".
+     * @returns {Promise<any>} An array of classifications for each input image or a single classification object if only one input image is provided.
+     */
+    // TODO: fix error below
```
Comment (kungfooman) [inline src/pipelines.js:898]: If you want to keep the todo, use `@todo` is JSDoc

File: src/pipelines.js [hunk 16]
```diff
@@ -925,7 +1149,19 @@ const TASK_ALIASES = {
...
+ * @async
+ * @function
+ * @param {string} task - The task to perform, e.g. "text-generation".
+ * @param {string} [model=null] - The name of the pre-trained model to use. If not specified, the default model for the task will be used.
+ * @param {object} [options] - Optional parameters for the pipeline.
+ * @param {function} [options.progress_callback=null] - A function to call with progress updates.
+ * @returns {Promise<Pipeline>} A Pipeline object for the specified task.
+ * @throws {Error} If an unsupported pipeline is requested.
+ */
+// TODO: fix error below
```
Comment (kungfooman) [inline src/pipelines.js:1164]: If you want to keep the todo, use `@todo` is JSDoc

File: src/processors.js [hunk 17]
```diff
@@ -277,6 +358,15 @@ class WhisperFeatureExtractor extends FeatureExtractor {
         return padded;
     }
 
+    /**
+     * Calculates the complex Short-Time Fourier Transform (STFT) of the given framed signal.
+     * 
+     * @param {Array<Array<number>>} frames - A 2D array representing the signal frames.
+     * @param {Array<number>} window - A 1D array representing the window to be applied to the frames.
```
Comment (kungfooman) [inline src/processors.js:365]: Currently we use two styles, `Array<number>` vs `number[]`, should we just stick to one?
```suggestion
     * @param {number[][]} frames - A 2D array representing the signal frames.
     * @param {number[]} window - A 1D array representing the window to be applied to the frames.
```

**Comment by kungfooman** [comment]
Great work, I looked through the PR once more and it just looks like minor issues, which can make it just a bit easier to read/maintain/code 👍 

I am impressed by all the work, I hope it doesn't feel too nitpicky 😅 🙈

File: src/pipelines.js [hunk 18]
```diff
@@ -232,27 +302,51 @@ class Text2TextGenerationPipeline extends Pipeline {
...
 
+        /**
+         * @type {any[]}
+         */
         let toReturn = this.tokenizer.batch_decode(outputTokenIds, {
             skip_special_tokens: true,
         });
         if (this._key !== null) {
-            toReturn = toReturn.map(text => {
+            toReturn = toReturn.map((text) => {
```
Comment (kungfooman) [inline src/pipelines.js:312]: (simply back to original)

```suggestion
            toReturn = toReturn.map(text => {
```

File: src/pipelines.js [hunk 19]
```diff
@@ -276,7 +373,7 @@ class TextGenerationPipeline extends Pipeline {
             let startText = texts[i].trim();
             let decoded = this.tokenizer.batch_decode(outTokens, {
                 skip_special_tokens: true,
-            }).map(x => {
+            }).map((x) => {
```
Comment (kungfooman) [inline src/pipelines.js:376]: (simply back to original to not change things besides JSDoc)

```suggestion
            }).map(x => {
```

File: src/pipelines.js [hunk 20]
```diff
@@ -626,7 +810,7 @@ class ImageToTextPipeline extends Pipeline {
             let output = (await this.model.generate(batch, generate_kwargs)).flat();
             let decoded = this.tokenizer.batch_decode(output, {
                 skip_special_tokens: true,
-            }).map(x => {
+            }).map((x) => {
```
Comment (kungfooman) [inline src/pipelines.js:813]: (back to original)

```suggestion
            }).map(x => {
```

File: src/pipelines.js [hunk 21]
```diff
@@ -736,7 +967,7 @@ class ObjectDetectionPipeline extends Pipeline {
         }
         images = await prepareImages(images);
 
-        let imageSizes = percentage ? null : images.map(x => [x.bitmap.width, x.bitmap.height]);
+        let imageSizes = percentage ? null : images.map((x) => [x.bitmap.width, x.bitmap.height]);
```
Comment (kungfooman) [inline src/pipelines.js:970]: (simply back to original)

```suggestion
        let imageSizes = percentage ? null : images.map(x => [x.bitmap.width, x.bitmap.height]);
```

File: src/tokenizers.js [hunk 22]
```diff
@@ -927,6 +1491,10 @@ class PreTrainedTokenizer extends Callable {
         return new this(tokenizerJSON, tokenizerConfig);
     }
 
+    /**
+     * @param {any} inputs
+     * @return {any}
+     */
```
Comment (kungfooman) [inline src/tokenizers.js:1497]: Merely to pass-through the type information so that it is not lost:

```suggestion
    /**
     * @param {T} inputs
     * @template T
     * @return {T}
     */
```

File: src/tokenizers.js [hunk 23]
```diff
@@ -987,7 +1555,10 @@ class PreTrainedTokenizer extends Callable {
         // Ensure it is less than model max length
         max_length = Math.min(max_length, this.model_max_length)
 
-        // TODO convert to tensor here?
+        // TODO: convert to tensor here?
+        /**
+         * @type {any}
+         */
```
Comment (kungfooman) [inline src/tokenizers.js:1561]: Converting to tensor looks like something to point out in an issue, if it is worth doing 
```suggestion
        /** @type {any[]} */
```

**Review by kungfooman — COMMENTED** [review]
Nearly takes an hour to go through everything 😅 

Great work, just pointing out a few things to keep the changes focused on JSDoc.

From my PoV this PR is ready after these few changes 💯

**Comment by chelouche9** [comment]
@xenova @kungfooman committed all changes. Thank you for the detailed review :)

Let me know if we need something else.

**Comment by xenova** [comment]
> @xenova @kungfooman committed all changes. Thank you for the detailed review :)
> 
> Let me know if we need something else.

I think that is it then!!! 😍 

I'm currently fixing a [bug](https://github.com/xenova/transformers.js/issues/63#issuecomment-1493397043) introduced by the latest version of optimum, which is quite important since it means no one can use their new models 👀 

I will get to merging this in the second that is fixed! Super excited! 🔥 Thanks to everyone for their hard work 🙏

**Comment by xenova** [comment]
@chelouche9 I'm almost done with my pass over everything (it's taking so long haha). I tried to push, but I apparently don't have write access to your branch. Could you add me as a collaborator?

**Comment by chelouche9** [comment]
@xenova now you have permissions.

**Comment by xenova** [comment]
> @xenova now you have permissions.

Thanks! I've pushed my changes. You and @kungfooman can do a quick skim over everything if you'd like... and then I'll merge into main 🔥

**Comment by chelouche9** [comment]
>Thanks! I've pushed my changes. You and @kungfooman can do a quick skim over everything if you'd like... and then I'll merge into main 🔥

I see you added some changes to the code except for jsdoc. For the annotation part looks good to me :)

**Comment by kungfooman** [comment]
> Also, I assume we have to run the build command each time before release? If so, then you're probably right that it's a good idea to remove dist from the repo. Perhaps @kungfooman can give some advice in that regard.

The only reason I like to commit build artifacts is to be able to see potential regressions, e.g. while rewriting a documentation system to keep track of the changes in the generated documentation files.

In such cases I just create a throw-away repo and commit as many files I want.

But I don't like all these artifact files in the "main repo", since I don't like large repositories. Cloning just takes longer and longer and it will probably also decrease the amount of contributors over time.

It's likely to result in merge conflicts aswell and I prefer not to spend time resolving merge conflicts in build artifacts.

I also like to use the file-search in VSCode and I don't want to put "dist" in "Exclude dirs" all the time, to filter for the SSOF (single source of truth) e.g. finding some specific types to improve.

TLDR: my suggestion is to remove /dist/ from repo
